### PR TITLE
Unit tests for KIP-714 client work

### PR DESCRIPTION
**Pull request by kirktrue**

### CODE REVIEW

**Review by philipnee — APPROVED** [review]
LGTM

File: clients/src/test/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterfaceTest.java [hunk 1]
```diff
@@ -102,47 +100,208 @@ public void testValidateClientInstanceId() {
...
 
-    private TelemetryMetric telemetryMetric(String name, long value) {
+    private TelemetryMetric newTelemetryMetric(String name, long value) {
         return new TelemetryMetric(name,
             MetricType.sum,
             value,
             "Description for " + name);
     }
+
+    private ProducerConfig newConfig() {
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterfaceTest.java:298]: maybe something like bootstrapTestProducerConfig?

**Review by philipnee — CHANGES_REQUESTED** [review]
Left a few comments.

File: clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java [hunk 2]
```diff
@@ -423,9 +408,26 @@ public static Uuid validateClientInstanceId(Uuid clientInstanceId) {
         return clientInstanceId.equals(Uuid.ZERO_UUID) ? Uuid.randomUuid() : clientInstanceId;
     }
 
-    public static int validatePushInteravlMs(Integer pushIntervalMs) {
-        // TODO: TELEMETRY_TODO: I don't think we'll need this, but maybe it's good to be paranoid?
-        return pushIntervalMs != null && pushIntervalMs > 0 ? pushIntervalMs : 10000;
+    public static int validatePushIntervalMs(Integer pushIntervalMs) {
```
[the remark this answers is not in the document]
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java]: Yes. Must have been a historical holdover. Cheerfully converted to an `int`.

File: clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java [hunk 3]
```diff
@@ -452,6 +454,32 @@ public static MetricType metricType(KafkaMetric kafkaMetric) {
...
     }
 
+    public List<TelemetryMetric> currentTelemetryMetrics(boolean deltaTemporality) {
+        return telemetryMetricsReporter.current().stream().map(kafkaMetric -> {
+            String name = kafkaMetric.metricName().name();
+            long value;
+
+            {
+                Object metricValue = kafkaMetric.metricValue();
+                double doubleValue = Double.parseDouble(metricValue.toString());
```
[the remark this answers is not in the document]
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java]: Possibly 🤷‍♂️ 

Not really sure what values we're getting out of the Kafka metrics layer or what the OTLP layer wants 😦

File: clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java [hunk 4]
```diff
@@ -452,6 +454,32 @@ public static MetricType metricType(KafkaMetric kafkaMetric) {
...
+                    Long previousValue = deltaValueStore.getAndSet(kafkaMetric.metricName(), value);
+                    value = previousValue != null ? value - previousValue : value;
+                }
+            }
+
+            MetricType metricType = metricType(kafkaMetric);
+            String description = kafkaMetric.metricName().description();
+
+            TelemetryMetric telemetryMetric = new TelemetryMetric(name, metricType, value, description);
+            log.debug("serialize including telemetry metric: {}", telemetryMetric);
```
[the remark this answers is not in the document]
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java]: Removed logging.

File: clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java [hunk 5]
```diff
@@ -486,20 +517,38 @@ public static String convertErrorToReason(Throwable error) {
...
+     */
+
+    static TelemetryState validateTransition(TelemetryState currState, TelemetryState newState) {
+        if (currState == null)
+            throw new IllegalTelemetryStateException("currState cannot be null");
+
+        if (newState == null)
+            throw new IllegalTelemetryStateException("newState cannot be null");
+
+        switch (currState) {
```
[the remark this answers is not in the document]
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java:523]: That's actually close to how the code was structured. I didn't like it at the time so I switched to how it is now. But I'm not married to any one approach.

Can we file that as a `TODO` outside of this PR?

File: clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java [hunk 6]
```diff
@@ -535,39 +584,71 @@ public static TelemetryState validateTransition(TelemetryState oldState, Telemet
...
 
-        throw new IllegalTelemetryStateException(String.format("Invalid transition from %s to %s", oldState, newState));
+        StringBuilder s = new StringBuilder();
+        s.append("Invalid telemetry state transition from ");
+        s.append(currState);
+        s.append(" to ");
+        s.append(newState);
+        s.append("; ");
+
+        if (allowableStates != null && allowableStates.length > 0) {
```
[the remark this answers is not in the document]
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java:621]: The return is up in the loop at the start of the method. The latter part is just formatting the error message. I put a little comment in and made the code a little more succinct.

LMK if there are further changes you'd like to see made here.

File: clients/src/test/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterfaceTest.java (hunk 1 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterfaceTest.java]: Changed to suggested name.

**Review by philipnee — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java [hunk 7]
```diff
@@ -452,6 +454,32 @@ public static MetricType metricType(KafkaMetric kafkaMetric) {
         }
     }
 
+    public List<TelemetryMetric> currentTelemetryMetrics(boolean deltaTemporality) {
+        return telemetryMetricsReporter.current().stream().map(kafkaMetric -> {
+            String name = kafkaMetric.metricName().name();
+            long value;
+
+            {
```
[the remark this answers is not in the document]
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/TelemetryManagementInterface.java]: Oh, that was just a little idiom I picked up to make everything in that section scoped so there weren't too many similar-looking variables in the main scope of the method. Kind of like a pseudo method.

Regardless, I removed it for clarity.

**Pull request merged**